firebase-firestore/src/main/java/com/google/firebase/firestore/util/ListenerRegistrationImpl.java

```diff
@@ -33,6 +58,119 @@
   /** The event listener for the query that raises events asynchronously. */
   private final ExecutorEventListener<ViewSnapshot> asyncEventListener;
 
+  static class CallbackList {
+    void run() {
+      for (Runnable callback : callbacks) {
+        if (callback != null) {
+          callback.run();
```

Comment:
https://github.com/firebase/firebase-android-sdk/blob/c627b48e40d3eedbb3a342bba61bf9a4b50088f9/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentReference.java#L478-L482

iiuc you also need to remove callbacks from the list as you're iterating over it or clear the list in onStop(), otherwise the callback can be called multiple times and also potentially it would leak memory, e.g. in the following scenario:

1. callback registered
2. activity stopped
3. fragment stopped
4. callback executes
5. activity restarts(via onRestart()+onStart()+onResume())
6. fragment starts again via onStart()(the same instance with a list full of callbacks)
7. now the next onStop will re-execute the callbacks from the last onStop()

Reply:
Yup; done. (Note: I think this also requires some sync to ensure clearing/adding don't happen at the same time.)